File: daily/Arrays/942.DI-String-Match.go
package daily

func diStringMatch(s string) []int {
    n := len(s)
    low, high := 0, n
    res := make([]int, n+1)
    for i := 0; i < n; i++ {
        if s[i] == 'I' {
            res[i] = low
            low++
        } else if s[i] == 'D' {
            res[i] = high
            high--
        }
    }

    if s[n-1] == 'I' {
        res[n] = high
    } else {
        res[n] = low
    }
    return res
}

File: daily/Arrays/942.DI-String-Match.py
from typing import List


class Solution:
    def diStringMatch(self, s: str) -> List[int]:
        n = len(s)
        low, high = 0, n
        res = [0] * (n + 1)
        for i, ch in enumerate(s):
            if ch == 'I':
                res[i] = low
                low += 1
            else:  # 'D'
                res[i] = high
                high -= 1
        res[n] = high if s[n - 1] == 'I' else low
        return res

